### PR TITLE
Porting fix-flapping patches from v21.6.4 by arutk

### PR DESCRIPTION
**Pull request by kmajzero**
Check superblock CRC before it is used

Superblock can be used during load of other sections, so we need to check
its CRC before other sections are loaded.

Signed-off-by: Robert Baldyga <robert.baldyga@intel.com>

Explicitly validate superblock after load

Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>

Remove ocf_metadata_probe_cores() implementation

This function must be fixed to work with metadata flapping. Until then
mark as not supported

Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>
Signed-off-by: Krzysztof Majzerowicz-Jaszcz <krzysztof.majzerowicz-jaszcz@intel.com>

Porting fix-flapping patches from v21.6.4 by arutk

Flapping fixes ported from v21.6.4 branch

Signed-off-by: Krzysztof Majzerowicz-Jaszcz <krzysztof.majzerowicz-jaszcz@intel.com>

### CODE REVIEW

**Comment by mmichal10** [comment]
Why have you squashed all the commits into one?

File: src/metadata/metadata.c [hunk 1]
```diff
@@ -1639,11 +1645,12 @@ void ocf_metadata_load_properties(ocf_volume_t volume,
 	int result;
 
 	OCF_DEBUG_TRACE(volume->cache);
-
+		
 	result = ocf_metadata_read_sb(volume->cache->owner, volume,
 			ocf_metadata_load_properties_cmpl, cmpl, priv);
-	if (result)
+	if (result)
```
Comment (arutk) [inline src/metadata/metadata.c:1651]: trailing whitespaces

File: src/metadata/metadata.c [hunk 2]
```diff
@@ -2003,35 +2010,9 @@ void ocf_metadata_probe(ocf_ctx_t ctx, ocf_volume_t volume,
...
-	context = env_vzalloc(sizeof(*context));
-	if (!context)
-		OCF_CMPL_RET(priv, -OCF_ERR_NO_MEM, 0);
-
-	context->cmpl = cmpl;
-	context->priv = priv;
-
-	ocf_metadata_query_cores(ctx, volume, uuids, uuids_count,
-			ocf_metadata_probe_cores_end, context);
+	OCF_CMPL_RET(priv, -OCF_ERR_NOT_SUPP, 0);
```
Comment (arutk) [inline src/metadata/metadata.c:2017]: this can be removed entirely now from OCF :)

Comment (arutk) [reply src/metadata/metadata.c:2017]: also there are some probe_cores functions still remaining in this file, which are unused even in current version

File: src/metadata/metadata_superblock.c [hunk 3]
```diff
@@ -98,42 +98,135 @@ static void ocf_metadata_store_segment(ocf_pipeline_t pipeline,
...
+{
+	struct ocf_metadata_context *context = priv;
+	struct ocf_metadata_ctrl *ctrl;
+	struct ocf_superblock_config *superblock;
+	int ret;
+
+	ctrl = (struct ocf_metadata_ctrl *)context->ctrl;
+	superblock = METADATA_MEM_POOL(ctrl, metadata_segment_sb_config);
+
+	ret = ocf_metadata_validate_superblock(NULL, superblock); // ocf_ctx_t here - how?
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:207]: context->cache->owner is the context

File: src/metadata/metadata_superblock.c [hunk 4]
```diff
@@ -98,42 +98,135 @@ static void ocf_metadata_store_segment(ocf_pipeline_t pipeline,
...
+	struct ocf_volume_uuid uuid;
+	ocf_volume_type_t volume_type;
+	ocf_core_t core;
+	ocf_core_id_t core_id;
+
+
+	ctrl = (struct ocf_metadata_ctrl *)context->ctrl;
 	sb_config = METADATA_MEM_POOL(ctrl, metadata_segment_sb_config);
 
 	if (sb_config->core_count > OCF_CORE_MAX) {
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:232]: these checks (lines 232-242) are now included in ocf_metadata_validate_superblock. If I recall corectly they were removed from here in 21.3 / 21.6 patches

File: src/metadata/metadata_superblock.c [hunk 5]
```diff
@@ -148,6 +241,19 @@ static void ocf_metadata_load_superblock_post(ocf_pipeline_t pipeline,
...
+		muuid = ocf_metadata_get_core_uuid(cache, core_id);
+		uuid.data = muuid->data;
+		uuid.size = muuid->size;
+
+		volume_type = ocf_ctx_get_volume_type(cache->owner,
+				core->conf_meta->type);
+
+		/* Initialize core volume */
+		ocf_volume_init(&core->volume, volume_type, &uuid, false);
+		core->has_volume = true;
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:254]: this is already done in _ocf_mngt_load_add_cores, you have effectively duplicated this part.

File: src/metadata/metadata_superblock.c [hunk 6]
```diff
@@ -230,9 +335,11 @@ struct ocf_pipeline_properties ocf_metadata_load_sb_pipeline_props = {
 	.steps = {
 		OCF_PL_STEP_FOREACH(ocf_metadata_store_segment,
 				ocf_metadata_load_sb_store_segment_args),
+		OCF_PL_STEP_ARG_INT(ocf_metadata_load_segment,
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:338]: analogous changes need to be applied to ocf_metadata_load_sb_recov_pipeline_props

File: src/metadata/metadata_superblock.c [hunk 7]
```diff
@@ -304,7 +411,7 @@ struct ocf_pipeline_properties ocf_metadata_load_sb_recov_pipeline_props = {
 				ocf_metadata_load_sb_store_segment_args),
 		OCF_PL_STEP_FOREACH(ocf_metadata_load_segment,
 				ocf_metadata_load_sb_recov_load_segment_args),
-		OCF_PL_STEP(ocf_metadata_check_crc_sb_config),
+	//	OCF_PL_STEP(ocf_metadata_check_crc_sb_config),
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:414]: please remove

File: src/metadata/metadata_superblock.c [hunk 8]
```diff
@@ -723,7 +789,7 @@ struct ocf_pipeline_properties ocf_metadata_sb_crc_recovery_pipeline_props = {
 	.priv_size = sizeof(struct ocf_metadata_context),
 	.finish = ocf_metadata_sb_crc_recovery_finish,
 	.steps = {
-		OCF_PL_STEP(ocf_metadata_check_crc_sb_config),
+		//OCF_PL_STEP(ocf_metadata_check_crc_sb_config),
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:792]: I'm completely unaware of this pipeline existence. I think it is not used anywhere? The only reference is in function ocf_metadata_sb_crc_recovery(), which doesn't appear to have any callers and is not in OCF API.

Comment (arutk) [reply src/metadata/metadata_superblock.c:792]: ... however I think the old implementation is correct, you should not remove the first step

**Review by arutk — CHANGES_REQUESTED** [review]

File: src/metadata/metadata.h [hunk 9]
```diff
@@ -150,7 +150,7 @@ void ocf_metadata_flush_mark(struct ocf_cache *cache, struct ocf_request *req,
  * @param queue - I/O queue to which metadata flush should be submitted
  * @param remaining - request remaining
  * @param complete - flushing request callback
- * @param context - context that will be passed into callback
+ * @param context - cvolume->ontext that will be passed into callback
```
Comment (arutk) [inline src/metadata/metadata.h:153]: can you please fix this comment parameters list?

**Review by arutk — CHANGES_REQUESTED** [review]

File: src/metadata/metadata_superblock.c [hunk 10]
```diff
@@ -303,12 +361,13 @@ struct ocf_pipeline_properties ocf_metadata_load_sb_recov_pipeline_props = {
 	.steps = {
 		OCF_PL_STEP_FOREACH(ocf_metadata_store_segment,
 				ocf_metadata_load_sb_store_segment_args),
+		OCF_PL_STEP_ARG_INT(ocf_metadata_load_segment,
+				metadata_segment_sb_config),
+		OCF_PL_STEP(_ocf_metadata_validate_superblock),
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:366]: this change belongs to previous commit, except that "_ocf_metadata_validate_superblock" would be ocf_metadata_check_crc_sb_config there.

File: src/metadata/metadata_superblock.c [hunk 11]
```diff
@@ -711,51 +729,6 @@ int ocf_metadata_read_sb(ocf_ctx_t ctx, ocf_volume_t volume,
...
-	OCF_DEBUG_TRACE(cache);
-
-	result = ocf_pipeline_create(&pipeline, cache,
-			&ocf_metadata_sb_crc_recovery_pipeline_props);
-	if (result)
-		OCF_CMPL_RET(priv, result);
-
-	context = ocf_pipeline_get_priv(pipeline);
-
-	context->cmpl = cmpl;
```
Comment (arutk) [inline src/metadata/metadata_superblock.c:749]: ideally removal of ocf_metadata_sb_crc_recovery would be in a separate commit

**Review by arutk — CHANGES_REQUESTED** [review]

File: src/metadata/metadata_superblock.c [hunk 12]
```diff
@@ -302,8 +303,8 @@ struct ocf_pipeline_properties ocf_metadata_load_sb_recov_pipeline_props = {
 	.steps = {
 		OCF_PL_STEP_FOREACH(ocf_metadata_store_segment,
 				ocf_metadata_load_sb_store_segment_args),
-		OCF_PL_STEP_FOREACH(ocf_metadata_load_segment,
-				ocf_metadata_load_sb_recov_load_segment_args),
+		OCF_PL_STEP_ARG_INT(ocf_metadata_load_segment,                                      
+				metadata_segment_sb_config),
```
Comment (robertbaldyga) [inline src/metadata/metadata_superblock.c:365]: I don't get what happened here. Anyway, `ocf_metadata_load_superblock_recovery()` is to be removed, so removing it in first commit will allow to avoid unnecessary changes in this code.

File: src/metadata/metadata.c [hunk 13]
```diff
@@ -1639,11 +1649,12 @@ void ocf_metadata_load_properties(ocf_volume_t volume,
 	int result;
 
 	OCF_DEBUG_TRACE(volume->cache);
-
+
```
Comment (robertbaldyga) [inline src/metadata/metadata.c]: Unnecessary whitespace.

File: src/metadata/metadata.c [hunk 14]
```diff
@@ -1639,11 +1649,12 @@ void ocf_metadata_load_properties(ocf_volume_t volume,
 	int result;
 
 	OCF_DEBUG_TRACE(volume->cache);
-
+	
 	result = ocf_metadata_read_sb(volume->cache->owner, volume,
 			ocf_metadata_load_properties_cmpl, cmpl, priv);
-	if (result)
+	if (result) {
```
Comment (robertbaldyga) [inline src/metadata/metadata.c]: No need to surround one-line block with the brackets.

File: src/metadata/metadata.c [hunk 15]
```diff
@@ -1649,12 +1646,11 @@ void ocf_metadata_load_properties(ocf_volume_t volume,
...
 	OCF_DEBUG_TRACE(volume->cache);
-	
+
 	result = ocf_metadata_read_sb(volume->cache->owner, volume,
 			ocf_metadata_load_properties_cmpl, cmpl, priv);
-	if (result) {
+	if (result)
 		OCF_CMPL_RET(priv, result, NULL);
-	}
 }
```
Comment (robertbaldyga) [inline src/metadata/metadata.c:1654]: Here you revert the change made in one of the previous commits. Please clean this up.

**Review by robertbaldyga — CHANGES_REQUESTED** [review]

**Review by robertbaldyga — APPROVED** [review]

**Review by arutk — APPROVED** [review]
